### PR TITLE
feat: add delegate multiaddrs to allow undialable nodes to connect to us

### LINKED ISSUE #8: Provide multiaddr for node handling pin request as `delegates`
> A list of temporary destination (retrievers) for the data. Returned by pinning service in a response for a pin request. 
> Both ends should attempt to preconnect to each other: 
> - Delegates should always preconnect to origins
> - Clients who initiate pin request and also have the pinned data in their own local datastore should preconnect to delegates

https://ipfs.github.io/pinning-services-api-spec/#section/Provider-hints

Figure out how to implement this. Currently we return an empty `delegates` array.

### PR DESCRIPTION
**Pull request by olizilla**
Provide delegate multiaddrs for a client to connect to for a pin add request (where possible) to allow nodes behind a NAT an ipfs node address for them to connect to, in case we can't connect to them.

- hold the request for up to ~20s in the lambda http handler while waiting for a worker node to pull the pin request from the queue and set its multiaddr on the pin record.
  - we hold it for less than 30s, as we're being called from a CF worker (timeout 30s) and via aws http gateway (timeout 30s)
- add PinTable class to update pin delegate info on pin record from pickup
- refactor `testIpfsApi` to be `findPublicMultiaddrs` and call as the first step of handling a pin request
- enable `assignPublicIp` in fargate task config so ipfs container has a public ip multiaddr
- update config to allow inbound ipv4 connections to 4001
- update ipfs config to set server profile so it doesn't advertise local ip addresses

Not tackled in this PR: failing pin request if no delegate assigned within timeout. We can't remove the message from the queue once it's sent. We _could_ set the pin as failed in the db, and have the worker read the pin state from the db at the start of handling it to check if it still needs processing.

fixes #8 

License: MIT

### CODE REVIEW

**Comment by olizilla** [comment]
Tested and working nice on PR build

```sh
❯ time curl -X POST -H "Authorization: Basic $CLUSTER_BASIC_AUTH_TOKEN" "https://pr134.pickup.dag.haus/pins/$cid" -sS | jq
{
  "replication_factor_min": -1,
  "replication_factor_max": -1,
  "name": "",
  "mode": "recursive",
  "shard_size": 0,
  "user_allocations": null,
  "expire_at": "0001-01-01T00:00:00Z",
  "metadata": {
    "delegates": [
      "/ip4/34.217.130.53/tcp/4001/p2p/12D3KooWSCSRyrNAP4ZcBHZAziWGwEtCtiZarJUDyu7LXGFUepdh",
      "/ip4/34.217.130.53/udp/4001/quic-v1/p2p/12D3KooWSCSRyrNAP4ZcBHZAziWGwEtCtiZarJUDyu7LXGFUepdh",
      "/ip4/34.217.130.53/udp/4001/quic/p2p/12D3KooWSCSRyrNAP4ZcBHZAziWGwEtCtiZarJUDyu7LXGFUepdh"
    ]
  },
  "pin_update": null,
  "origins": [],
  "cid": "bafkreif6idhkvmjld6qu3c6u33ubjx4amrzzhpdoywcf4yhmwib6yzlu6m",
  "type": "pin",
  "allocations": [],
  "max_depth": -1,
  "reference": null,
  "timestamp": "2023-03-21T13:28:56.519Z"
}

________________________________________________________
Executed in    2.93 secs      fish           external
   usr time   49.73 millis    0.14 millis   49.58 millis
   sys time   14.41 millis    1.62 millis   12.79 millis

```

**Comment by olizilla** [comment]
GET /pins/:cid response
```sh
❯ time curl -X GET -H "Authorization: Basic $CLUSTER_BASIC_AUTH_TOKEN" "https://pr134.pickup.dag.haus/pins/$cid" -sS | jq
{
  "cid": "bafkreif6idhkvmjld6qu3c6u33ubjx4amrzzhpdoywcf4yhmwib6yzlu6m",
  "name": "",
  "allocations": [],
  "origins": [],
  "created": "2023-03-21T13:28:56.519Z",
  "metadata": {
    "delegates": [
      "/ip4/34.217.130.53/tcp/4001/p2p/12D3KooWSCSRyrNAP4ZcBHZAziWGwEtCtiZarJUDyu7LXGFUepdh",
      "/ip4/34.217.130.53/udp/4001/quic-v1/p2p/12D3KooWSCSRyrNAP4ZcBHZAziWGwEtCtiZarJUDyu7LXGFUepdh",
      "/ip4/34.217.130.53/udp/4001/quic/p2p/12D3KooWSCSRyrNAP4ZcBHZAziWGwEtCtiZarJUDyu7LXGFUepdh"
    ]
  },
  "peer_map": {
    "12D3KooWArSKMUUeLk3z2m5LKyb9wGyFL1BtWCT7Gq7Apoo77PUR": {
      "peername": "elastic-ipfs",
      "ipfs_peer_id": "bafzbeigjqot6fm3i3yv37wiyybsfblrlsmib7bzlbnkpjxde6fw6b4fvei",
      "ipfs_peer_addresses": [
        "/dns4/staging.peer.ipfs-elastic-provider-aws.com/tcp/3000/ws/p2p/bafzbeigjqot6fm3i3yv37wiyybsfblrlsmib7bzlbnkpjxde6fw6b4fvei"
      ],
      "status": "pin_queued",
      "timestamp": "2023-03-21T13:31:51.141Z",
      "error": "",
      "attempt_count": 0,
      "priority_pin": false
    }
  }
}

________________________________________________________
Executed in    1.27 secs      fish           external
   usr time   48.95 millis    0.17 millis   48.77 millis
   sys time   17.43 millis    1.87 millis   15.56 millis

```

**Comment by olizilla** [comment]
Fixed UDP & TCP mapping, and can now connect to both tcp and quic addrs

```sh
❯ ipfs swarm connect /ip4/34.222.160.153/tcp/4001/p2p/12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU
connect 12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU success

❯ ipfs swarm peers | grep 12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU
/ip4/34.222.160.153/tcp/4001/p2p/12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU

❯ ipfs swarm disconnect /p2p/12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU
disconnect 12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU success

❯ ipfs swarm connect /ip4/34.222.160.153/udp/4001/quic-v1/p2p/12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU
connect 12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU success

❯ ipfs swarm peers | grep 12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU
/ip4/34.222.160.153/udp/4001/quic-v1/p2p/12D3KooWFa6GgfAomSQbNLGkW2Dh3LJwuPwUyCs1AE6WCA2bM4VU
```

File: api/package.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "version": "1.0.0",
   "type": "module",
   "scripts": {
-    "test": "ava"
+    "test": "SST_APP=NOSUCH SST_STAGE=NOSUCH LOG_LEVEL=fatal ava"
```
Comment (vasco-santos) [inline api/package.json:6]: nit: better to have `.env.tpl` and local copy of it?

File: api/basic/add-pin.ts [hunk 2]
```diff
@@ -101,27 +104,58 @@ export async function handler (event: APIGatewayProxyEventV2, context: Context):
...
+  }
+  // wait for delegates to be set on Pin, or return the Pin withtout if we don't get them in time
+  try {
+    const pinWithDelegates = await retry(async () => await findDelegates({ cid, dynamo, table }), {
+      minTimeout: 2000,
+      retries: 4,
+      factor: 1.66 // spread 4 retries over ~20s, see 'Real Solution' on https://www.wolframalpha.com/input?i=Sum%5B1000*x%5Ek,+%7Bk,+0,+5%7D%5D+%3D+27+*+1000
+    })
+    return toAddPinResponse(pinWithDelegates, origins)
+  } catch (error) {
```
Comment (alanshaw) [inline api/basic/add-pin.ts:125]: Log the error?

File: pickup/test/pickup.test.js [hunk 3]
```diff
@@ -18,6 +19,28 @@ test.after(async t => {
   await t.context.shutDownDockers()
 })
 
+test.only('add delegates to a pin record', async t => {
```
Comment (alanshaw) [inline pickup/test/pickup.test.js:22]:
```suggestion
test('add delegates to a pin record', async t => {
```

**Review by alanshaw — APPROVED** [review]

**Review by vasco-santos — APPROVED** [review]

File: pickup/ipfs/ipfs-config.sh [hunk 4]
```diff
@@ -16,9 +22,6 @@ ipfs config --json Swarm.DisableBandwidthMetrics true
 ipfs config --json Swarm.ConnMgr.HighWater 100
 ipfs config --json Swarm.ConnMgr.LowWater 50
 
-# no MDNS plz
-ipfs config --json Discovery.MDNS.Enabled false
```
[the remark this answers is not in the document]
Comment (olizilla) [reply pickup/ipfs/ipfs-config.sh:20]: is set when we enable the server profile

File: pickup/lib/pickup.js [hunk 5]
```diff
@@ -50,13 +58,16 @@ export function createPickupFromEnv (env = process.env) {
...
  */
-export function createPickup ({ sqsPoller, carFetcher, s3Uploader }) {
+export function createPickup ({ sqsPoller, carFetcher, s3Uploader, pinTable }) {
   /**
    * @param {import('squiss-ts').Message} msg
    */
   async function messageHandler (msg) {
     const { cid, origins, key } = msg.body
+    const delegates = await carFetcher.findPublicMultiaddrs()
+    await pinTable.addDelegates({ cid, delegates })
```
[the remark this answers is not in the document]
Comment (olizilla) [reply pickup/lib/pickup.js:70]: the set size will typically be `3`

**Pull request merged**